cranelift/codegen/src/opts/bitops.isle

```diff
@@ -92,3 +92,17 @@
 (rule (simplify (bxor ty x (iconst ty k)))
   (if-let -1 (i64_sextend_imm64 ty k))
   (bnot ty x))
+
+;; bmask(input) = !(((input | ((!input) + 1)) >> 63) - 1)
+(rule (simplify (bnot ty
+                      (isub ty
+                            (ushr ty
+                                  (bor ty
+                                        input
+                                        (iadd ty
+                                              (bnot ty input)
+                                              (iconst ty (u64_from_imm64 1))))
+                                  (iconst ty (u64_from_imm64 shift_amt)))
+                            (iconst ty (u64_from_imm64 1)))))
+      (if-let $true (u64_eq shift_amt (u64_sub (ty_bits ty) 1)))
+      (bmask ty input))
```

Comment:
Is this actually correct? I couldn't convince myself, so I asked Z3 to find an `x` for which `bmask(x)` and `!(((x | ((!x) + 1)) >> (ty_bits - 1)) - 1)` were not equal (that is, I asked it to find a counterexample):

```lisp
(declare-const x (_ BitVec 32))

(assert (! (= (ite (= x #x00000000)
                   #x00000000
                   #xffffffff)
              ;; !(((x | ((!x) + 1)) >> 31) - 1)
              (bvnot (bvsub (bvlshr (bvor x
                                          (bvadd (bvnot x)
                                                 #x00000001))
                                    #x0000001f)
                            #x00000001)))))

(check-sat)
(get-model)
```

It reports `sat`, meaning that it found a counterexample, and it says the counterexample is zero:

```lisp
sat
(
  (define-fun x () (_ BitVec 32)
    #x00000000)
)
```

So unless I'm missing something and/or translated the expression incorrectly, I think this optimization is incorrect?

Reply:
Then again, there is an explicit test for 0 in the run tests, so I must have done something wrong...

Reply:
The SMT counterexample is confusing me too:
- `x1 := (bvnot x)` with `x := 0` -> `0xffffffff`
- `x2 := (bvadd x1 1)` -> `0`
- `x3 := (bvor x2 x)` -> `(bvor 0 0)` -> `0`
- `x4 := (bvlshr x3 31)` -> `0`
- `x5 := (bvsub x4 1)` -> `0xffffffff`
- `x6 := (bvnot x5)` -> `0`

and on the LHS for the if-then-else we have `(ite (= 0 0) 0 0xffffffff)` -> `0`. Seems like the optimization did what it was supposed to? I too am confused why we get SAT here.

Reply:
That Z3 expression looks like a correct translation of that pattern to my eyes, so I'm confused too.

I think the rule seems correct:
- `!input + 1` is equivalent to `-input` in two's-complement.
- 0 is the only value which, when negated, leaves the sign bit clear. (`INT_MIN` leaves the sign bit set; all other cases flip the sign bit.)
- Shifting the sign bit down after `x | -x` then gives us 0 if the input was 0, and 1 otherwise.
- Subtracting 1 means we get -1 if the input was 0, and 0 otherwise.
- Inverting the result should then give us the right answer.

Reply:
(@afonso360 a comment to this effect in the source would be very useful for the future!)

Reply:
Ah, and after posting I see @jameysharp 's comment too. Both of us arrived at the same reasoning more or less so I'm fairly confident here :-)

Reply:
We can extract out more broadly-applicable rules that should compose nicely into this one.

Rewrite both of these patterns to `(ineg ty input)`:
- `(iadd ty (bnot ty input) (iconst ty (u64_from_imm64 1)))`
- `(bnot ty (isub ty input (iconst ty (u64_from_imm64 1))))`

In a 64-bit word, `-(x >> 63)`, where right-shift is unsigned, should I think be equivalent to `x >> 63` with a signed right-shift. Or in general, rewrite negation of unsigned shifts by N-1 bits in N-bit words.

Then this rule only needs to match `(x | -x) >> (N-1)`, with a signed right-shift.

That should make all these rules more broadly applicable as well as (somewhat) easier to understand.

`x | -x` leaves the trailing 0s alone and sets the remaining bits to 1. (I found the inverse of this pattern in "Hacker's Delight", on page 12.) I don't think there's a useful rule we can extract from that pattern alone.

Reply:
The problem with the smtlib version was the use of `!`, which is for annotating expressions: switching to `not` fixes the problem:
```
(declare-const x (_ BitVec 32))                                                  
                                                                                 
(assert (not (= (ite (= x #x00000000)                                            
                   #x00000000                                                    
                   #xffffffff)                                                   
              ;; !(((x | ((!x) + 1)) >> 31) - 1)                                 
              (bvnot (bvsub (bvlshr (bvor x                                      
                                          (bvadd (bvnot x)                       
                                                 #x00000001))                    
                                    #x0000001f)                                  
                            #x00000001)))))                                      
                                                                                 
(check-sat)                                                                      
(get-model)  
```
produces the following output:
```
% z3 -smt2 test.smt
unsat
(error "line 14 column 10: model is not available")
```